azurelinuxagent/ga/remoteaccess.py

```diff
@@ -0,0 +1,163 @@
+                                                    ExtensionStatus, \
+                                                    ExtensionSubStatus, \
+                                                    VMStatus, ExtHandler, \
+                                                    get_properties, \
+                                                    set_properties
+from azurelinuxagent.common.protocol.metadata import MetadataProtocol
+from azurelinuxagent.common.utils.cryptutil import CryptUtil
+from azurelinuxagent.common.utils.flexible_version import FlexibleVersion
+from azurelinuxagent.common.utils.processutil import capture_from_process
+from azurelinuxagent.common.protocol import get_protocol_util
+from azurelinuxagent.common.version import AGENT_NAME, CURRENT_VERSION
+from azurelinuxagent.common.osutil import get_osutil
+
+REMOTE_USR_EXPIRATION_FORMAT = "%a, %d %b %Y %H:%M:%S %Z"
+DATE_FORMAT = "%Y-%m-%d"
+TRANSPORT_PRIVATE_CERT = "TransportPrivate.pem"
+REMOTE_ACCESS_ACCOUNT_COMMENT = "JIT Account"
+MAX_TRY_ATTEMPT = 5
+FAILED_ATTEMPT_THROTTLE = 1
+
+def get_remote_access_handler():
+    return RemoteAccessHandler()
+
+class RemoteAccessHandler(object):
+    def __init__(self):
+        self.protocol_util = get_protocol_util()
+        self.protocol = None
+        self.os_util = None
+        self.cryptUtil = CryptUtil(conf.get_openssl_cmd())
+        self.remote_access = None
+
+    def run(self):
+        try:
+            if self.os_util is None:
+                self.os_util = get_osutil()
+            if self.os_util.jit_enabled:
+                logger.verbose("Handle remote access updates")
+                self.protocol = self.protocol_util.get_protocol()
+                self.protocol.client.update_goal_state(True)
+                self.protocol.client.update_remote_access_conf(self.protocol.client.goal_state)
+                if self.remote_access is None or self.remote_access.incarnation != self.protocol.client.remote_access.incarnation:
+                    self.remote_access = self.protocol.client.remote_access
+                    self.handle_remote_access()
+            else:
+                logger.info("Non JIT enabled client.")
+        except Exception as e:
+            msg = u"Exception processing remote access handler: {0}".format(
+                ustr(e))
+            logger.warn(msg)
+            logger.info(str(e))
+            add_event(AGENT_NAME,
+                      version=CURRENT_VERSION,
+                      op=WALAEventOperation.HandleRemoteAccess,
+                      is_success=False,
+                      message=msg)
+            return     
+
+    def handle_remote_access(self):
+        logger.verbose("Entered handle_remote_access")        
+        if self.remote_access is not None:
+            # Get JIT user accounts.
+            all_users = self.os_util.get_users()
+            jit_users = set()
+            for usr in all_users:
+                if usr[4] == REMOTE_ACCESS_ACCOUNT_COMMENT:
```

Comment:
This seems fragile

Reply:
This is mostly meant for the delete scenario that is not yet implemented but also helps to ignore non-JIT accounts to compare against in the add scenario since it's not needed.  It's not meant to prevent accounts from sticking around if a user modifies intentionally modifies it.  With sudo privileges there are many things a user can do to defeat JIT cleanup mechanisms (create another account, for example) but the AM team will eventually add functionality to scan for these extra accounts and raise flags in security compliance reports if users do this.

---
In reply to: [192195605](https://github.com/Azure/WALinuxAgent/pull/1171#discussion_r192195605) [](ancestors = 192195605)

Reply:
I would prefer a method to make it clear what this test is meant to achieve.

```python
  def is_jit_account(usr):
     return usr[4] == REMOTE_ACCESS_ACCOUNT_COMMENT
```

If there is a better way to make this check that would be desirable. I know the names are random, but do they follow a pattern? Do they always have a specific prefix?

---
In reply to: [192198874](https://github.com/Azure/WALinuxAgent/pull/1171#discussion_r192198874) [](ancestors = 192198874,192195605)

Reply:
I can change to a regex on the username, but I thought this would be more reliable than something like TA[0-9]*[a-z]*[0-9]{2}_ 

---
In reply to: [192216362](https://github.com/Azure/WALinuxAgent/pull/1171#discussion_r192216362) [](ancestors = 192216362,192198874,192195605)

Reply:
Second Chris's comment here, please move check to a method.

Reply:
moved

---
In reply to: [194026384](https://github.com/Azure/WALinuxAgent/pull/1171#discussion_r194026384) [](ancestors = 194026384)